azkaban-common/src/main/java/azkaban/executor/container/watch/AzPodStatusExtractor.java

```diff
@@ -0,0 +1,527 @@
+    // Phase must be failed.
+      if (podPhase != PodPhase.Failed) {
+        logger.debug("InitFailed is false and phase is not failed");
+        return false;
+    }
+    // Initalized conditions should not be true
+    if (initializedConditionStatus.isPresent() &&
+        initializedConditionStatus.get() == PodConditionStatus.True) {
+      logger.debug("InitFailed is failed as initialized conditions is not present or true");
+      return false;
+    }
+
+    // There must be at least 1 failed init container
+    List<V1ContainerStatus> initContainerStatuses = v1PodStatus.getInitContainerStatuses();
+    if (initContainerStatuses == null || initContainerStatuses.isEmpty()) {
+      logger.debug("InitFailed is false as init container status is null or empty");
+      return false;
+    }
+    boolean anyContainerFailed = initContainerStatuses.stream().anyMatch(status ->
+        status.getState().getTerminated() != null &&
+            (status.getState().getTerminated().getExitCode() == null ||
+                status.getState().getTerminated().getExitCode() != 0));
+    if (!anyContainerFailed) {
+      logger.debug("InitFailed is false as as all init container are terminated with exit code 0");
+      return false;
+    }
+    logger.debug("InitFailed is true");
+    return true;
+  }
+
+  /**
+   * Evaluate if event can be classified as status {@link AzPodStatus.AZ_POD_APP_FAILURE}
+   * @return true if the classification succeeds, false otherwise.
+   */
+  private boolean checkForAzPodAppFailure() {
+    // Phase must be failed.
+    if (podPhase != PodPhase.Failed) {
+      logger.debug("AppFailed is false and phase is not failed");
+      return false;
+    }
+    // Initialized condition should  be true
+    if (!initializedConditionStatus.isPresent() ||
+        initializedConditionStatus.get() != PodConditionStatus.True) {
+      logger.debug("AppFailed is failed as initialized conditions is not present or not true");
+      return false;
+    }
+    // There must be at least 1 failed app container
+    List<V1ContainerStatus> containerStatuses = v1PodStatus.getInitContainerStatuses();
+    if (containerStatuses == null || containerStatuses.isEmpty()) {
+      logger.debug("AppFailed is false as container status is null or empty");
+      return false;
+    }
+    boolean anyContainerFailed = containerStatuses.stream().anyMatch(status ->
+        status.getState().getTerminated() != null &&
+            (status.getState().getTerminated().getExitCode() == null ||
+                status.getState().getTerminated().getExitCode() != 0));
+    if (!anyContainerFailed) {
+      logger.debug("AppFailed is false as as no container terminated with non-zero exit code");
+      return false;
+    }
+    logger.debug("AppFailed is true");
+    return true;
+  }
+
+  /**
+   * Return the {@link AzPodStatus} derived from given Pod watch event.
+   * @return
+   */
+  public AzPodStatus createAzPodStatus() {
```

Comment:
This is looking a bit inefficient to me. I dont have a solution but inorder to run checkForAzkPodAppFailure, in the worst case, all the conditions must be checked before.
Is there a way to avoid that?
Perhaps as a TODO for future?

Reply:
Actually, I had the opposite in mind. The `checkAz*` methods are currently implemented in a manner where for every raw `Watch<V1Pod>` event exactly one of the `checkAz...` methods should return true. 
Running the raw event through all the methods will serve as a good validation for this logic (e.g we can add a warning in case more than 1 method returns true).
All of these methods will run very quickly, as they include simple point lookup of variables and in some cases traversals of arrays with 5-10 elements.

Reply:
In that case, how about just precompute the results in constructor and cache them for access?

Reply:
That's definitely worth considering. How about I revisit this part in a subsequent PR? (will create a task for this).

Reply:
Sure, can you please create a ticket to track this internally and resolve this issue?

Reply:
Added an internal tracking ticket.

Reply:
@sshardool if >1 check methods return true, which status should this method return?